src/engines/TrainElectricEngine.cpp

```diff
@@ -24,42 +23,353 @@ namespace godot {
+        ClassDB::bind_method(
+                D_METHOD("get_power_cable_steam_pressure"), &TrainElectricEngine::get_power_cable_steam_pressure);
+        ADD_PROPERTY(
+                PropertyInfo(Variant::FLOAT, "power/power_cable/steam_pressure"), "set_power_cable_steam_pressure",
+                "get_power_cable_steam_pressure");
     }
 
     void TrainElectricEngine::_do_fetch_state_from_mover(TMoverParameters *mover, Dictionary &state) {
         TrainEngine::_do_fetch_state_from_mover(mover, state);
 
         state["compressor_enabled"] = mover->CompressorFlag;
         state["compressor_allowed"] = mover->CompressorAllow;
-
         state["converter_enabled"] = mover->ConverterFlag;
         state["converted_allowed"] = mover->ConverterAllow;
         state["converter_time_to_start"] = mover->ConverterStartDelayTimer;
+        state["power_source"] = static_cast<int>(mover->EnginePowerSource.SourceType);
+        state["accumulator/recharge_source"] = static_cast<int>(mover->EnginePowerSource.RAccumulator.RechargeSource);
+        state["current_collector/max_voltage"] = mover->EnginePowerSource.MaxVoltage;
+        state["current_collector/max_current"] = mover->EnginePowerSource.MaxCurrent;
+        state["current_collector/max_collector_lifting"] = mover->EnginePowerSource.CollectorParameters.MaxH;
+        state["current_collector/min_collector_lifting"] = mover->EnginePowerSource.CollectorParameters.MinH;
+        state["current_collector/csw"] = mover->EnginePowerSource.CollectorParameters.CSW;
+        state["current_collector/min_main_switch_voltage"] = mover->EnginePowerSource.CollectorParameters.MinV;
+        state["current_collector/min_pantograph_tank_pressure"] = mover->EnginePowerSource.CollectorParameters.MinPress;
+        state["current_collector/max_pantograph_tank_pressure"] = mover->EnginePowerSource.CollectorParameters.MaxPress;
+        state["current_collector/overvoltage_relay"] = mover->EnginePowerSource.CollectorParameters.OVP;
+        state["current_collector/required_main_switch_voltage"] = mover->EnginePowerSource.CollectorParameters.InsetV;
+        state["transducer/input_voltage"] = mover->EnginePowerSource.Transducer.InputVoltage;
+        state["power_cable/source"] = static_cast<int>(mover->EnginePowerSource.RPowerCable.PowerTrans);
+        state["power_cable/steam_pressure"] = mover->EnginePowerSource.RPowerCable.SteamPressure;
     }
 
     void TrainElectricEngine::_do_update_internal_mover(TMoverParameters *mover) {
         TrainEngine::_do_update_internal_mover(mover);
 
         // FIXME: test data
-        mover->EnginePowerSource.SourceType = TPowerSource::CurrentCollector;
-
+        mover->EnginePowerSource.SourceType = static_cast<TPowerSource>(power_source);
         mover->CompressorSwitch(compressor_switch_pressed);
         mover->ConverterSwitch(converter_switch_pressed);
+        switch (mover->EnginePowerSource.SourceType) {
```

Comment:
W tej metodzie aktualizujemy movera z propsów node'a. Powinno być `switch (power_source)`.

Wiem że wyżej zostaje ustawiony SourceType, ale teraz działanie tego bloku jest zależny od tego czy SourceType zostanie nastawione przed czy po czy w ogóle. Tj w skrajnym przypadku, gdyby usutawienie PowerSource bylo po bloku switch, to parametry będą ustawiane o jedną ramkę za późno. 

Choć teraz kod jest poprawny, to taką konstrukcje w kodzie traktowałbym jako swoistą pułapkę.

Reply:
Ta ale wtedy musielibyśmy dwa razy statycznie castowac power_source do enuma

Reply:
To problem? Bo IMO lepiej jest mieć mniej (lub nie mieć) pułapek w źródłach. 
A to podwójne castowanie kompilator może zoptymalizować tak, że różnicy raczej będzie.

W tym warunku "boli" uzależnienie bloku kodu od stanu obiektu docelowego, a nie źródłowego.